File: src/main/java/org/academy/api/common/profiler/AcademyProfiler.java
package org.academy.api.common.profiler;

import java.util.List;
import java.util.function.Supplier;

public final class AcademyProfiler {
    private static final long SNAPSHOT_TTL_NANOS = 250_000_000L;

    private static volatile long cachedSnapshotAt = 0L;
    private static volatile ProfilerSnapshot cachedSnapshot = null;

    private AcademyProfiler() {
    }

    public static void push(String name) {
        ZoneProfiler.push(name);
    }

    public static void pop() {
        ZoneProfiler.pop();
    }

    public static void popPush(String name) {
        ZoneProfiler.popPush(name);
    }

    public static <T> T zone(String name, Supplier<T> block) {
        ZoneProfiler.push(name);
        try {
            return block.get();
        } finally {
            ZoneProfiler.pop();
        }
    }

    public static void runZone(String name, Runnable block) {
        ZoneProfiler.push(name);
        try {
            block.run();
        } finally {
            ZoneProfiler.pop();
        }
    }

    public static void incrementCounter(String name, int amount) {
        ZoneProfiler.incrementCounter(name, amount);
    }

    public static void incrementCounter(String name) {
        ZoneProfiler.incrementCounter(name, 1);
    }

    public static boolean isCapturingZones() {
        return ZoneProfiler.isEnabled();
    }

    public static void startZoneCapture() {
        ZoneProfiler.setEnabled(true);
        invalidateSnapshot();
    }

    public static void stopZoneCapture() {
        ZoneProfiler.setEnabled(false);
        invalidateSnapshot();
    }

    public static void resetZones() {
        ZoneProfiler.reset();
        invalidateSnapshot();
    }

    public static boolean isSampling() {
        return ProfilerSampler.isRunning();
    }

    public static boolean isSamplingPaused() {
        return ProfilerSampler.isPaused();
    }

    public static void startSampling() {
        ProfilerSampler.start(1000L);
        invalidateSnapshot();
    }

    public static void startSampling(long intervalMicros) {
        ProfilerSampler.start(intervalMicros);
        invalidateSnapshot();
    }

    public static void stopSampling() {
        ProfilerSampler.stop();
        invalidateSnapshot();
    }

    public static void pauseSampling() {
        ProfilerSampler.pause();
        invalidateSnapshot();
    }

    public static void resumeSampling() {
        ProfilerSampler.resume();
        invalidateSnapshot();
    }

    public static void resetSampling() {
        ProfilerSampler.reset();
        invalidateSnapshot();
    }

    public static ProfilerSampler.ThreadRef registerThread(Thread thread) {
        return ProfilerSampler.registerThread(thread);
    }

    public static void unregisterThread(Thread thread) {
        ProfilerSampler.unregisterThread(thread.threadId());
    }

    public static void setThreadEnabled(long threadId, boolean enabled) {
        ProfilerSampler.setThreadEnabled(threadId, enabled);
    }

    public static List<ProfilerSampler.ThreadRef> samplerThreads() {
        return ProfilerSampler.threadRefs();
    }

    private static void invalidateSnapshot() {
        cachedSnapshot = null;
        cachedSnapshotAt = 0L;
    }

    public static ProfilerSnapshot snapshot() {
        long now = System.nanoTime();
        ProfilerSnapshot snap = cachedSnapshot;
        if (snap != null && now - cachedSnapshotAt < SNAPSHOT_TTL_NANOS) {
            return snap;
        }
        ProfilerSnapshot fresh = buildSnapshot();
        cachedSnapshot = fresh;
        cachedSnapshotAt = System.nanoTime();
        return fresh;
    }

    private static ProfilerSnapshot buildSnapshot() {
        return new ProfilerSnapshot(
                ZoneProfiler.snapshot(),
                ProfilerSampler.hasData() ? ProfilerSampler.snapshot() : null,
                FrameStats.snapshot(),
                ZoneProfiler.isEnabled(),
                ProfilerSampler.isRunning(),
                ProfilerSampler.isPaused()
        );
    }
}

File: src/main/kotlin/org/academy/api/common/profiler/AcademyProfiler.kt
package org.academy.api.common.profiler

/**
 * AcademyCraft 性能分析总入口（门面）。
 *
 * 客户端 / 服务端 / 命令共用。数据采集完全自包含，不依赖 MC Profiler / Tracy / ImGui。
 */
object AcademyProfiler {
    // ---------- Zones（push/pop 插桩） ----------

    @JvmStatic
    fun push(name: String) = ZoneProfiler.push(name)

    @JvmStatic
    fun pop() = ZoneProfiler.pop()

    @JvmStatic
    fun popPush(name: String) = ZoneProfiler.popPush(name)

    @JvmStatic
    fun <T> zone(name: String, block: () -> T): T {
        ZoneProfiler.push(name)
        try {
            return block()
        } finally {
            ZoneProfiler.pop()
        }
    }

    /** Java 友好的 zone 便捷方法。 */
    @JvmStatic
    fun runZone(name: String, block: Runnable) {
        ZoneProfiler.push(name)
        try {
            block.run()
        } finally {
            ZoneProfiler.pop()
        }
    }

    @JvmStatic
    fun incrementCounter(name: String, amount: Int) = ZoneProfiler.incrementCounter(name, amount)

    @JvmStatic
    fun incrementCounter(name: String) = ZoneProfiler.incrementCounter(name, 1)

    @JvmStatic
    fun isCapturingZones(): Boolean = ZoneProfiler.enabled

    @JvmStatic
    fun startZoneCapture() {
        ZoneProfiler.setEnabled(true)
        invalidateSnapshot()
    }

    @JvmStatic
    fun stopZoneCapture() {
        ZoneProfiler.setEnabled(false)
        invalidateSnapshot()
    }

    @JvmStatic
    fun resetZones() {
        ZoneProfiler.reset()
        invalidateSnapshot()
    }

    // ---------- Sampler（采样式全量剖析） ----------

    @JvmStatic
    fun isSampling(): Boolean = ProfilerSampler.isRunning

    @JvmStatic
    fun isSamplingPaused(): Boolean = ProfilerSampler.isPaused

    @JvmStatic
    fun startSampling() {
        ProfilerSampler.start(1000L)
        invalidateSnapshot()
    }

    @JvmStatic
    fun startSampling(intervalMicros: Long) {
        ProfilerSampler.start(intervalMicros)
        invalidateSnapshot()
    }

    @JvmStatic
    fun stopSampling() {
        ProfilerSampler.stop()
        invalidateSnapshot()
    }

    @JvmStatic
    fun pauseSampling() {
        ProfilerSampler.pause()
        invalidateSnapshot()
    }

    @JvmStatic
    fun resumeSampling() {
        ProfilerSampler.resume()
        invalidateSnapshot()
    }

    @JvmStatic
    fun resetSampling() {
        ProfilerSampler.reset()
        invalidateSnapshot()
    }

    @JvmStatic
    fun registerThread(thread: Thread) = ProfilerSampler.registerThread(thread)

    @JvmStatic
    fun unregisterThread(thread: Thread) = ProfilerSampler.unregisterThread(thread.id)

    @JvmStatic
    fun setThreadEnabled(threadId: Long, enabled: Boolean) = ProfilerSampler.setThreadEnabled(threadId, enabled)

    @JvmStatic
    fun samplerThreads(): List<ProfilerSampler.ThreadRef> = ProfilerSampler.threadRefs()

    // ---------- 快照 ----------

    @Volatile
    private var cachedSnapshotAt = 0L

    @Volatile
    private var cachedSnapshot: ProfilerSnapshot? = null

    private const val SNAPSHOT_TTL_NANOS = 250_000_000L

    private fun invalidateSnapshot() {
        cachedSnapshot = null
        cachedSnapshotAt = 0L
    }

    @JvmStatic
    fun snapshot(): ProfilerSnapshot {
        val now = System.nanoTime()
        cachedSnapshot?.let {
            if (now - cachedSnapshotAt < SNAPSHOT_TTL_NANOS) return it
        }
        val fresh = buildSnapshot()
        cachedSnapshot = fresh
        cachedSnapshotAt = System.nanoTime()
        return fresh
    }

    private fun buildSnapshot(): ProfilerSnapshot {
        return ProfilerSnapshot(
            zones = ZoneProfiler.snapshot(),
            sampler = if (ProfilerSampler.hasData) ProfilerSampler.snapshot() else null,
            frame = FrameStats.snapshot(),
            zonesEnabled = ZoneProfiler.enabled,
            sampling = ProfilerSampler.isRunning,
            samplingPaused = ProfilerSampler.isPaused,
        )
    }
}
